diff -ruN a/crossly.hs b/crossly.hs
--- a/crossly.hs
+++ b/crossly.hs
@@ -1671,11 +1671,7 @@
 #define FORWARD 127
 #define REDUCING 126
 
-#if defined(__M2__)
-#define CELL_SIZE sizeof(unsigned)
-#else
 #define CELL_SIZE 1
-#endif
 
 unsigned* mem;
 unsigned* altmem;
diff -ruN a/marginally.hs b/marginally.hs
--- a/marginally.hs
+++ b/marginally.hs
@@ -1500,11 +1500,7 @@
 
 #define TOP 16777216
 
-#if defined(__M2__)
-#define CELL_SIZE sizeof(unsigned)
-#else
 #define CELL_SIZE 1
-#endif
 
 
 unsigned* mem;
diff -ruN a/methodically.hs b/methodically.hs
--- a/methodically.hs
+++ b/methodically.hs
@@ -775,7 +775,9 @@
   (   adt
   <|> classDecl
   <|> instDecl
   <|> res "ffi" *> (addFFI <$> wantString <*> var <*> (res "::" *> _type))
+  <|> res "foreign" *> res "import" *> var *>
+    (addFFI <$> wantString <*> var <*> (res "::" *> _type))
   <|> res "export" *> (addExport <$> wantString <*> var)
   <|> addDefs <$> def
   <|> fixity *> pure id
@@ -1436,11 +1438,7 @@
 
 #define TOP 16777216
 
-#if defined(__M2__)
-#define CELL_SIZE sizeof(unsigned)
-#else
 #define CELL_SIZE 1
-#endif
 
 
 unsigned* mem;
diff -ruN a/precisely.hs b/precisely.hs
--- a/precisely.hs
+++ b/precisely.hs
@@ -1782,11 +1782,7 @@
 #define FORWARD 127
 #define REDUCING 126
 
-#if defined(__M2__)
-#define CELL_SIZE sizeof(unsigned)
-#else
 #define CELL_SIZE 1
-#endif
 
 unsigned* mem;
 unsigned* altmem;
diff -ruN a/virtually.hs b/virtually.hs
--- a/virtually.hs
+++ b/virtually.hs
@@ -1286,11 +1286,7 @@
 
 #define TOP 16777216
 
-#if defined(__M2__)
-#define CELL_SIZE sizeof(unsigned)
-#else
 #define CELL_SIZE 1
-#endif
 
 unsigned* mem;
 unsigned* altmem;
diff -ruN a/vm.c b/vm.c
--- a/vm.c
+++ b/vm.c
@@ -28,11 +28,7 @@
 #define TABMAX    1024
 #define BUFMAX    1048576
 
-#if defined(__M2__)
-	#define CELL_SIZE sizeof(unsigned)
-#else
-	#define CELL_SIZE 1
-#endif
+#define CELL_SIZE 1
 
 unsigned* mem;
 unsigned* altmem;
@@ -193,8 +189,18 @@
 void reset(unsigned root)
 {
+	unsigned prompt;
+	unsigned start;
+	unsigned halted;
+	unsigned tail;
+	unsigned whole;
 	root_memo = root;
+	prompt = app('0', '?');
+	start = app(root, prompt);
+	halted = app(start, '.');
+	tail = app('T', '1');
+	whole = app(halted, tail);
 	sp = spTop;
-	sp[0] = app(app(app(root, app('0', '?')), '.'), app('T', '1'));
+	sp[0] = whole;
 }
 
 void load(char* filename)
@@ -773,7 +779,8 @@
 			str = buf;
 			rts_init();
 			sp = spTop;
-			spTop[0] = app(app(starting_address, '?'), '.');
+			starting_address = app(starting_address, '?');
+			spTop[0] = app(starting_address, '.');
 			option_index = option_index + 1;
 		}
 		else if(match(argv[option_index], "-pb") || match(argv[option_index], "--parse-buffer"))
