--- tccelf.c
+++ tccelf.c
@@ -2111,6 +2111,10 @@ static int set_sec_sizes(TCCState *s1)
     int textrel = 0;
     int file_type = s1->output_type;
 
+    s = have_section(s1, ".init");
+    if (s) s->sh_flags |= SHF_ALLOC | SHF_EXECINSTR;
+    s = have_section(s1, ".fini");
+    if (s) s->sh_flags |= SHF_ALLOC | SHF_EXECINSTR;
     /* Allocate strings for section names */
     for(i = 1; i < s1->nb_sections; i++) {
         s = s1->sections[i];
