diff --git a/include/stddef.h b/include/stddef.h
index 8f885fcd..5326ef38 100644
--- a/include/stddef.h
+++ b/include/stddef.h
@@ -33 +33 @@ typedef unsigned long long int uint64_t;
-void *alloca(size_t size);
+typedef union { long double ld; long long ll; } max_align_t; void *alloca(size_t size);
diff --git a/libtcc.c b/libtcc.c
index a9950319..cdaf8fc8 100644
--- a/libtcc.c
+++ b/libtcc.c
@@ -884 +884 @@ LIBTCCAPI TCCState *tcc_new(void)
-#if defined(TCC_MUSL)
+#if defined(TCC_MUSL) || defined(TCC_MES_LIBC)
diff --git a/tccelf.c b/tccelf.c
index 2b19f77c..b880d113 100644
--- a/tccelf.c
+++ b/tccelf.c
@@ -2241 +2241 @@ static int elf_output_file(TCCState *s1, const char *filename)
-        fill_got(s1);
+        { fill_got(s1); relocate_plt(s1); }
diff --git a/tccgen.c b/tccgen.c
index 487dbf18..7855a3c1 100644
--- a/tccgen.c
+++ b/tccgen.c
@@ -4154,0 +4155,7 @@ static int post_type(CType *type, AttributeDef *ad, int storage, int td)
+        if (tok != ']') {
+            while (tok == TOK_STATIC
+                   || tok == TOK_CONST1 || tok == TOK_CONST2 || tok == TOK_CONST3
+                   || tok == TOK_VOLATILE1 || tok == TOK_VOLATILE2 || tok == TOK_VOLATILE3
+                   || tok == TOK_RESTRICT1 || tok == TOK_RESTRICT2 || tok == TOK_RESTRICT3)
+                next();
+        }
diff --git a/tcctools.c b/tcctools.c
index 990fbccc..5d93c12c 100644
--- a/tcctools.c
+++ b/tcctools.c
@@ -127 +127 @@ ST_FUNC int tcc_tool_ar(TCCState *s1, int argc, char **argv)
-    if (!i_obj)  // i_obj implies also i_lib. we require both.
+    if (!i_lib)
@@ -138,0 +139,6 @@ ST_FUNC int tcc_tool_ar(TCCState *s1, int argc, char **argv)
+    if (!i_obj) {
+        fwrite("!<arch>\n", 8, 1, fh);
+        ret = 0;
+        goto the_end;
+    }
+
