--- a/libtcc.c
+++ b/libtcc.c
@@ -884,6 +884,9 @@
     s->warn_implicit_function_declaration = 1;
     s->warn_discarded_qualifiers = 1;
     s->ms_extensions = 1;
+#if BOOTSTRAP
+    s->static_link = 1;
+#endif
     s->unwind_tables = 1;
 
 #ifdef CHAR_IS_UNSIGNED
--- a/tccelf.c
+++ b/tccelf.c
@@ -3128,7 +3128,7 @@
 	}
         /* Perform relocation to GOT or PLT entries */
         if (file_type == TCC_OUTPUT_EXE && s1->static_link)
-            fill_got(s1);
+            { fill_got(s1); relocate_plt(s1); }
         else if (s1->got)
             fill_local_got_entries(s1);
 
--- a/tccgen.c
+++ b/tccgen.c
@@ -3082,6 +3082,7 @@
             if (op != '-')
                 goto op_err;
             vpush_type_size(pointed_type(&vtop[-1].type), &align);
+            if (!(vtop[-1].type.t & VT_UNSIGNED)) gen_cast_s(VT_PTRDIFF_T);
             vtop->type.t &= ~VT_UNSIGNED;
             vrott(3);
             gen_opic(op);
@@ -3105,6 +3106,7 @@
 #endif
             type1 = vtop[-1].type;
             vpush_type_size(pointed_type(&vtop[-1].type), &align);
+            if (!(vtop[-1].type.t & VT_UNSIGNED)) gen_cast_s(VT_PTRDIFF_T);
             vtop->type.t &= ~VT_UNSIGNED;
             gen_op('*');
 #ifdef CONFIG_TCC_BCHECK
--- a/include/stdarg.h
+++ b/include/stdarg.h
@@ -1,11 +1,19 @@
 #ifndef _STDARG_H
 #define _STDARG_H

+#ifdef __HCC__
+typedef char *va_list;
+#define va_start(ap, last) ((void)((ap) = (char *)(&(last)) + sizeof(void *)))
+#define va_arg(ap, type) (type)(((long *)((ap) = ((ap) + sizeof(void *))))[-1])
+#define va_end(ap) ((void)((ap) = 0))
+#define va_copy(d, s) ((d) = (s))
+#else
 typedef __builtin_va_list va_list;
 #define va_start __builtin_va_start
 #define va_arg __builtin_va_arg
 #define va_copy __builtin_va_copy
 #define va_end __builtin_va_end
+#endif
 
 /* fix a buggy dependency on GCC in libio.h */
 typedef va_list __gnuc_va_list;
