diff --git a/precisely.hs b/precisely.hs
index 554858b..828304b 100644
--- a/precisely.hs
+++ b/precisely.hs
@@ -90,7 +90,8 @@ instance Integral Word where
 instance Eq Word where (==) = wordEq
 instance Ord Word where (<=) = wordLE
 
-data Word64 = Word64 Word Word deriving Eq
+data Word64 = Word64 Word Word
+instance Eq Word64 where Word64 a b == Word64 c d = a == c && b == d
 instance Ring Word64 where
   Word64 a b + Word64 c d = uncurry Word64 $ word64Add a b c d
   Word64 a b - Word64 c d = uncurry Word64 $ word64Sub a b c d
@@ -106,7 +107,8 @@ instance Ord Word64 where
     | True = b <= d
 
 -- Multiprecision arithmetic.
-data Integer = Integer Bool [Word] deriving Eq
+data Integer = Integer Bool [Word]
+instance Eq Integer where Integer xsgn xs == Integer ysgn ys = xsgn == ysgn && xs == ys
 instance Ring Integer where
   Integer xsgn xs + Integer ysgn ys
     | xsgn == ysgn = Integer xsgn $ mpAdd xs ys
@@ -212,6 +214,10 @@ class Monad m where
 liftA2 f x y = f <$> x <*> y
 (>>) f g = f >>= \_ -> g
 class Eq a where (==) :: a -> a -> Bool
+instance Eq Bool where
+  True == True = True
+  False == False = True
+  _ == _ = False
 instance Eq Int where (==) = intEq
 instance Eq Char where (==) = charEq
 ($) f x = f x
@@ -429,13 +435,13 @@ foldrWithKey f = go where
 toAscList = foldrWithKey (\k x xs -> (k,x):xs) []
 
 -- Syntax tree.
-data Type = TC String | TV String | TAp Type Type deriving Eq
+data Type = TC String | TV String | TAp Type Type
 arr a b = TAp (TAp (TC "->") a) b
 data Extra = Basic Int | Const Integer | ChrCon Char | StrCon String
 data Pat = PatLit Extra | PatVar String (Maybe Pat) | PatCon String [Pat]
 data Ast = E Extra | V String | A Ast Ast | L String Ast | Pa [([Pat], Ast)] | Ca Ast [(Pat, Ast)] | Proof Pred
 data Constr = Constr String [(String, Type)]
-data Pred = Pred String Type deriving Eq
+data Pred = Pred String Type
 data Qual = Qual [Pred] Type
 noQual = Qual []
 
@@ -444,6 +450,14 @@ typeVars = \case
   TV v -> [v]
   TAp x y -> typeVars x `union` typeVars y
 
+instance Eq Type where
+  (TC s) == (TC t) = s == t
+  (TV s) == (TV t) = s == t
+  (TAp a b) == (TAp c d) = a == c && b == d
+  _ == _ = False
+
+instance Eq Pred where (Pred s a) == (Pred t b) = s == t && a == b
+
 data Instance = Instance
   -- Type, e.g. Int for Eq Int.
   Type
@@ -813,7 +827,12 @@ addLets ls x = foldr triangle x components where
     insLams vs t = foldr L t vs
     in foldr (\(x:xt) t -> A (L x t) $ maybeFix x $ insLams xt $ suball $ maybe undefined id $ lookup x ls) (suball expr) tnames
 
-data Assoc = NAssoc | LAssoc | RAssoc deriving Eq
+data Assoc = NAssoc | LAssoc | RAssoc
+instance Eq Assoc where
+  NAssoc == NAssoc = True
+  LAssoc == LAssoc = True
+  RAssoc == RAssoc = True
+  _ == _ = False
 precOf s precTab = maybe 5 fst $ mlookup s precTab
 assocOf s precTab = maybe LAssoc snd $ mlookup s precTab
 
