diff --git a/.github/workflows/build.yml b/.github/workflows/build.yml
deleted file mode 100644
index e2ed28e..0000000
--- a/.github/workflows/build.yml
+++ /dev/null
@@ -1,29 +0,0 @@
-name: Build and test
-
-on: [push, pull_request]
-
-jobs:
-  build-test:
-    runs-on: ubuntu-latest
-    steps:
-    - uses: actions/checkout@v2.3.1
-    - uses: cachix/install-nix-action@v12
-      with:
-        nix_path: nixpkgs=channel:nixos-unstable
-    - name: Build
-      run: nix-build -A blynn-compiler
-      timeout-minutes: 20
-    - name: Check hash of raw
-      run: echo "f803e4ca8cf281ec971f76a4920767242aa8427850ffbef85f69c4d01f5f3cdd result/share/raw" | sha256sum -c
-    - run: cp result/bin/vm vm && cp result/share/raw raw && cp result/bin/precisely precisely
-    - run: ./precisely test/hello.hs hello.c && gcc -O2 -o hello hello.c && ./hello > hello.txt
-    - name: Check hash of hello world test
-      run: echo "4dca0fd5f424a31b03ab807cbae77eb32bf2d089eed1cee154b3afed458de0dc  hello.txt" | sha256sum -c
-    - name: Upload artifacts
-      uses: actions/upload-artifact@v2
-      with:
-        name: build result
-        path: |
-          raw
-          vm
-          precisely
diff --git a/.gitmodules b/.gitmodules
deleted file mode 100644
index b67fb1a..0000000
--- a/.gitmodules
+++ /dev/null
@@ -1,3 +0,0 @@
-[submodule "M2libc"]
-	path = M2libc
-	url = https://github.com/oriansj/M2libc.git
diff --git a/default.nix b/default.nix
deleted file mode 100644
index 4fb8073..0000000
--- a/default.nix
+++ /dev/null
@@ -1,8 +0,0 @@
-{ sources ? import ./nix/sources.nix { }
-, pkgs ? import sources.nixpkgs { }
-}:
-let
-  # Nixpkgs extended with bootstrappable related packages
-  bootstrappable-pkgs = import ./pkgs { };
-in
-bootstrappable-pkgs
diff --git a/go.sh b/go.sh
index 166ba3c..7b8f836 100755
--- a/go.sh
+++ b/go.sh
@@ -8,38 +8,15 @@ mkdir -p bin
 # Place to put generated source files
 mkdir -p generated
 
-# compile pack_blobs.c
-M2-Planet --architecture x86 \
-	-f M2libc/sys/types.h \
-	-f M2libc/stddef.h \
-	-f M2libc/x86/Linux/unistd.h \
-	-f M2libc/stdlib.c \
-	-f M2libc/x86/Linux/fcntl.h \
-	-f M2libc/stdio.c \
-	-f M2libc/bootstrappable.c \
-	-f pack_blobs.c \
-	--debug \
-	-o bin/pack_blobs.M1
+M2_ARCH=${M2_ARCH:-x86}
+M2_OS=${M2_OS:-Linux}
 
-# Create dwarf stubs for pack_blobs
-blood-elf -f bin/pack_blobs.M1 --little-endian --entry _start -o bin/pack_blobs-footer.M1
+compile_m2() {
+	M2-Mesoplanet --operating-system "$M2_OS" --architecture "$M2_ARCH" -f "$1" -o "$2"
+	chmod 555 "$2"
+}
 
-# Convert to hex2 linker format
-M1 -f M2libc/x86/x86_defs.M1 \
-	-f M2libc/x86/libc-full.M1 \
-	-f bin/pack_blobs.M1 \
-	-f bin/pack_blobs-footer.M1 \
-	--little-endian \
-	--architecture x86 \
-	-o bin/pack_blobs.hex2
-
-# Link into final static binary
-hex2 -f M2libc/x86/ELF-x86-debug.hex2 \
-	-f bin/pack_blobs.hex2 \
-	--little-endian \
-	--architecture x86 \
-	--base-address 0x8048000 \
-	-o bin/pack_blobs
+compile_m2 pack_blobs.c bin/pack_blobs
 
 # Build blobs
 ./bin/pack_blobs -f blob/parenthetically.source -o generated/parenthetically
@@ -47,38 +24,7 @@ hex2 -f M2libc/x86/ELF-x86-debug.hex2 \
 ./bin/pack_blobs -f blob/practically.source -o generated/practically
 ./bin/pack_blobs -f blob/singularity.source -o generated/singularity_blob
 
-# Compile to assembly vm.c
-M2-Planet --architecture x86 \
-	-f M2libc/sys/types.h \
-	-f M2libc/stddef.h \
-	-f M2libc/x86/Linux/unistd.h \
-	-f M2libc/stdlib.c \
-	-f M2libc/x86/Linux/fcntl.h \
-	-f M2libc/stdio.c \
-	-f M2libc/bootstrappable.c \
-	-f vm.c \
-	--debug \
-	-o bin/vm.M1
-
-# Create dwarf stubs for vm
-blood-elf -f bin/vm.M1 --little-endian --entry _start -o bin/vm-footer.M1
-
-# Convert to hex2 linker format
-M1 -f M2libc/x86/x86_defs.M1 \
-	-f M2libc/x86/libc-full.M1 \
-	-f bin/vm.M1 \
-	-f bin/vm-footer.M1 \
-	--little-endian \
-	--architecture x86 \
-	-o bin/vm.hex2
-
-# Link into final static binary
-hex2 -f M2libc/x86/ELF-x86-debug.hex2 \
-	-f bin/vm.hex2 \
-	--little-endian \
-	--architecture x86 \
-	--base-address 0x8048000 \
-	-o bin/vm
+compile_m2 vm.c bin/vm
 
 # Generate raw file needed
 ./bin/vm --raw blob/root -pb bootstrap -lf generated/parenthetically -o bin/raw_l
@@ -119,126 +65,14 @@ hex2 -f M2libc/x86/ELF-x86-debug.hex2 \
 # Make virtually
 ./bin/vm -f virtually.hs --foreign 2 --raw generated/uniquely_raw.txt --rts_c run -o generated/virtually_raw.txt
 
-# Make marginally
 ./bin/vm -f marginally.hs --foreign 2 --raw generated/virtually_raw.txt --rts_c run -o generated/marginally.c
-M2-Planet --architecture x86 \
-	-f M2libc/sys/types.h \
-	-f M2libc/stddef.h \
-	-f M2libc/x86/Linux/unistd.h \
-	-f M2libc/stdlib.c \
-	-f M2libc/x86/Linux/fcntl.h \
-	-f M2libc/stdio.c \
-	-f M2libc/bootstrappable.c \
-	-f generated/marginally.c \
-	--debug \
-	-o bin/marginally.M1
+compile_m2 generated/marginally.c bin/marginally
 
-blood-elf -f bin/marginally.M1 --little-endian --entry _start -o bin/marginally-footer.M1
-
-M1 -f M2libc/x86/x86_defs.M1 \
-	-f M2libc/x86/libc-full.M1 \
-	-f bin/marginally.M1 \
-	-f bin/marginally-footer.M1 \
-	--little-endian \
-	--architecture x86 \
-	-o bin/marginally.hex2
-
-hex2 -f M2libc/x86/ELF-x86-debug.hex2 \
-	-f bin/marginally.hex2 \
-	--little-endian \
-	--architecture x86 \
-	--base-address 0x8048000 \
-	-o bin/marginally
-
-# Make methodically
 ./bin/marginally methodically.hs generated/methodically.c
-M2-Planet --architecture x86 \
-	-f M2libc/sys/types.h \
-	-f M2libc/stddef.h \
-	-f M2libc/x86/Linux/unistd.h \
-	-f M2libc/stdlib.c \
-	-f M2libc/x86/Linux/fcntl.h \
-	-f M2libc/stdio.c \
-	-f M2libc/bootstrappable.c \
-	-f generated/methodically.c \
-	--debug \
-	-o bin/methodically.M1
+compile_m2 generated/methodically.c bin/methodically
 
-blood-elf -f bin/methodically.M1 --little-endian --entry _start -o bin/methodically-footer.M1
-
-M1 -f M2libc/x86/x86_defs.M1 \
-	-f M2libc/x86/libc-full.M1 \
-	-f bin/methodically.M1 \
-	-f bin/methodically-footer.M1 \
-	--little-endian \
-	--architecture x86 \
-	-o bin/methodically.hex2
-
-hex2 -f M2libc/x86/ELF-x86-debug.hex2 \
-	-f bin/methodically.hex2 \
-	--little-endian \
-	--architecture x86 \
-	--base-address 0x8048000 \
-	-o bin/methodically
-
-# Make crossly
 ./bin/methodically crossly.hs generated/crossly.c
-M2-Planet --architecture x86 \
-	-f M2libc/sys/types.h \
-	-f M2libc/stddef.h \
-	-f M2libc/x86/Linux/unistd.h \
-	-f M2libc/stdlib.c \
-	-f M2libc/x86/Linux/fcntl.h \
-	-f M2libc/stdio.c \
-	-f M2libc/bootstrappable.c \
-	-f generated/crossly.c \
-	--debug \
-	-o bin/crossly.M1
+compile_m2 generated/crossly.c bin/crossly
 
-blood-elf -f bin/crossly.M1 --little-endian --entry _start -o bin/crossly-footer.M1
-
-M1 -f M2libc/x86/x86_defs.M1 \
-	-f M2libc/x86/libc-full.M1 \
-	-f bin/crossly.M1 \
-	-f bin/crossly-footer.M1 \
-	--little-endian \
-	--architecture x86 \
-	-o bin/crossly.hex2
-
-hex2 -f M2libc/x86/ELF-x86-debug.hex2 \
-	-f bin/crossly.hex2 \
-	--little-endian \
-	--architecture x86 \
-	--base-address 0x8048000 \
-	-o bin/crossly
-
-# Make precisely
 ./bin/crossly precisely.hs generated/precisely.c
-M2-Planet --architecture x86 \
-	-f M2libc/sys/types.h \
-	-f M2libc/stddef.h \
-	-f M2libc/x86/Linux/unistd.h \
-	-f M2libc/stdlib.c \
-	-f M2libc/x86/Linux/fcntl.h \
-	-f M2libc/stdio.c \
-	-f M2libc/bootstrappable.c \
-	-f generated/precisely.c \
-	--debug \
-	-o bin/precisely.M1
-
-blood-elf -f bin/precisely.M1 --little-endian --entry _start -o bin/precisely-footer.M1
-
-M1 -f M2libc/x86/x86_defs.M1 \
-	-f M2libc/x86/libc-full.M1 \
-	-f bin/precisely.M1 \
-	-f bin/precisely-footer.M1 \
-	--little-endian \
-	--architecture x86 \
-	-o bin/precisely.hex2
-
-hex2 -f M2libc/x86/ELF-x86-debug.hex2 \
-	-f bin/precisely.hex2 \
-	--little-endian \
-	--architecture x86 \
-	--base-address 0x8048000 \
-	-o bin/precisely
+compile_m2 generated/precisely.c bin/precisely
diff --git a/nix/sources.json b/nix/sources.json
deleted file mode 100644
index f90b8cf..0000000
--- a/nix/sources.json
+++ /dev/null
@@ -1,26 +0,0 @@
-{
-    "niv": {
-        "branch": "master",
-        "description": "Easy dependency management for Nix projects",
-        "homepage": "https://github.com/nmattia/niv",
-        "owner": "nmattia",
-        "repo": "niv",
-        "rev": "ba57d5a29b4e0f2085917010380ef3ddc3cf380f",
-        "sha256": "1kpsvc53x821cmjg1khvp1nz7906gczq8mp83664cr15h94sh8i4",
-        "type": "tarball",
-        "url": "https://github.com/nmattia/niv/archive/ba57d5a29b4e0f2085917010380ef3ddc3cf380f.tar.gz",
-        "url_template": "https://github.com/<owner>/<repo>/archive/<rev>.tar.gz"
-    },
-    "nixpkgs": {
-        "branch": "release-20.03",
-        "description": "Nix Packages collection",
-        "homepage": "",
-        "owner": "NixOS",
-        "repo": "nixpkgs",
-        "rev": "0f8a31b992b6892dd075bac8533955a9b92ac9df",
-        "sha256": "11013clgxalivmmh3m4cglg51sxms5kgbqh2bqdhcngra3cqnlsy",
-        "type": "tarball",
-        "url": "https://github.com/NixOS/nixpkgs/archive/0f8a31b992b6892dd075bac8533955a9b92ac9df.tar.gz",
-        "url_template": "https://github.com/<owner>/<repo>/archive/<rev>.tar.gz"
-    }
-}
diff --git a/nix/sources.nix b/nix/sources.nix
deleted file mode 100644
index b796fff..0000000
--- a/nix/sources.nix
+++ /dev/null
@@ -1,171 +0,0 @@
-# This file has been generated by Niv.
-
-let
-
-  #
-  # The fetchers. fetch_<type> fetches specs of type <type>.
-  #
-
-  fetch_file = pkgs: name: spec:
-    let
-      name' = sanitizeName name + "-src";
-    in
-      if spec.builtin or true then
-        builtins_fetchurl { inherit (spec) url sha256; name = name'; }
-      else
-        pkgs.fetchurl { inherit (spec) url sha256; name = name'; };
-
-  fetch_tarball = pkgs: name: spec:
-    let
-      name' = sanitizeName name + "-src";
-    in
-      if spec.builtin or true then
-        builtins_fetchTarball { name = name'; inherit (spec) url sha256; }
-      else
-        pkgs.fetchzip { name = name'; inherit (spec) url sha256; };
-
-  fetch_git = name: spec:
-    let
-      ref =
-        if spec ? ref then spec.ref else
-          if spec ? branch then "refs/heads/${spec.branch}" else
-            if spec ? tag then "refs/tags/${spec.tag}" else
-              abort "In git source '${name}': Please specify `ref`, `tag` or `branch`!";
-    in
-      builtins.fetchGit { url = spec.repo; inherit (spec) rev; inherit ref; };
-
-  fetch_local = spec: spec.path;
-
-  fetch_builtin-tarball = name: throw
-    ''[${name}] The niv type "builtin-tarball" is deprecated. You should instead use `builtin = true`.
-        $ niv modify ${name} -a type=tarball -a builtin=true'';
-
-  fetch_builtin-url = name: throw
-    ''[${name}] The niv type "builtin-url" will soon be deprecated. You should instead use `builtin = true`.
-        $ niv modify ${name} -a type=file -a builtin=true'';
-
-  #
-  # Various helpers
-  #
-
-  # https://github.com/NixOS/nixpkgs/pull/83241/files#diff-c6f540a4f3bfa4b0e8b6bafd4cd54e8bR695
-  sanitizeName = name:
-    (
-      concatMapStrings (s: if builtins.isList s then "-" else s)
-        (
-          builtins.split "[^[:alnum:]+._?=-]+"
-            ((x: builtins.elemAt (builtins.match "\\.*(.*)" x) 0) name)
-        )
-    );
-
-  # The set of packages used when specs are fetched using non-builtins.
-  mkPkgs = sources: system:
-    let
-      sourcesNixpkgs =
-        import (builtins_fetchTarball { inherit (sources.nixpkgs) url sha256; }) { inherit system; };
-      hasNixpkgsPath = builtins.any (x: x.prefix == "nixpkgs") builtins.nixPath;
-      hasThisAsNixpkgsPath = <nixpkgs> == ./.;
-    in
-      if builtins.hasAttr "nixpkgs" sources
-      then sourcesNixpkgs
-      else if hasNixpkgsPath && ! hasThisAsNixpkgsPath then
-        import <nixpkgs> {}
-      else
-        abort
-          ''
-            Please specify either <nixpkgs> (through -I or NIX_PATH=nixpkgs=...) or
-            add a package called "nixpkgs" to your sources.json.
-          '';
-
-  # The actual fetching function.
-  fetch = pkgs: name: spec:
-
-    if ! builtins.hasAttr "type" spec then
-      abort "ERROR: niv spec ${name} does not have a 'type' attribute"
-    else if spec.type == "file" then fetch_file pkgs name spec
-    else if spec.type == "tarball" then fetch_tarball pkgs name spec
-    else if spec.type == "git" then fetch_git name spec
-    else if spec.type == "local" then fetch_local spec
-    else if spec.type == "builtin-tarball" then fetch_builtin-tarball name
-    else if spec.type == "builtin-url" then fetch_builtin-url name
-    else
-      abort "ERROR: niv spec ${name} has unknown type ${builtins.toJSON spec.type}";
-
-  # If the environment variable NIV_OVERRIDE_${name} is set, then use
-  # the path directly as opposed to the fetched source.
-  replace = name: drv:
-    let
-      saneName = stringAsChars (c: if isNull (builtins.match "[a-zA-Z0-9]" c) then "_" else c) name;
-      ersatz = builtins.getEnv "NIV_OVERRIDE_${saneName}";
-    in
-      if ersatz == "" then drv else ersatz;
-
-  # Ports of functions for older nix versions
-
-  # a Nix version of mapAttrs if the built-in doesn't exist
-  mapAttrs = builtins.mapAttrs or (
-    f: set: with builtins;
-    listToAttrs (map (attr: { name = attr; value = f attr set.${attr}; }) (attrNames set))
-  );
-
-  # https://github.com/NixOS/nixpkgs/blob/0258808f5744ca980b9a1f24fe0b1e6f0fecee9c/lib/lists.nix#L295
-  range = first: last: if first > last then [] else builtins.genList (n: first + n) (last - first + 1);
-
-  # https://github.com/NixOS/nixpkgs/blob/0258808f5744ca980b9a1f24fe0b1e6f0fecee9c/lib/strings.nix#L257
-  stringToCharacters = s: map (p: builtins.substring p 1 s) (range 0 (builtins.stringLength s - 1));
-
-  # https://github.com/NixOS/nixpkgs/blob/0258808f5744ca980b9a1f24fe0b1e6f0fecee9c/lib/strings.nix#L269
-  stringAsChars = f: s: concatStrings (map f (stringToCharacters s));
-  concatMapStrings = f: list: concatStrings (map f list);
-  concatStrings = builtins.concatStringsSep "";
-
-  # https://github.com/NixOS/nixpkgs/blob/8a9f58a375c401b96da862d969f66429def1d118/lib/attrsets.nix#L331
-  optionalAttrs = cond: as: if cond then as else {};
-
-  # fetchTarball version that is compatible between all the versions of Nix
-  builtins_fetchTarball = { url, name ? null, sha256 }@attrs:
-    let
-      inherit (builtins) lessThan nixVersion fetchTarball;
-    in
-      if lessThan nixVersion "1.12" then
-        fetchTarball ({ inherit url; } // (optionalAttrs (!isNull name) { inherit name; }))
-      else
-        fetchTarball attrs;
-
-  # fetchurl version that is compatible between all the versions of Nix
-  builtins_fetchurl = { url, name ? null, sha256 }@attrs:
-    let
-      inherit (builtins) lessThan nixVersion fetchurl;
-    in
-      if lessThan nixVersion "1.12" then
-        fetchurl ({ inherit url; } // (optionalAttrs (!isNull name) { inherit name; }))
-      else
-        fetchurl attrs;
-
-  # Create the final "sources" from the config
-  mkSources = config:
-    mapAttrs (
-      name: spec:
-        if builtins.hasAttr "outPath" spec
-        then abort
-          "The values in sources.json should not have an 'outPath' attribute"
-        else
-          spec // { outPath = replace name (fetch config.pkgs name spec); }
-    ) config.sources;
-
-  # The "config" used by the fetchers
-  mkConfig =
-    { sourcesFile ? if builtins.pathExists ./sources.json then ./sources.json else null
-    , sources ? if isNull sourcesFile then {} else builtins.fromJSON (builtins.readFile sourcesFile)
-    , system ? builtins.currentSystem
-    , pkgs ? mkPkgs sources system
-    }: rec {
-      # The sources, i.e. the attribute set of spec name to spec
-      inherit sources;
-
-      # The "pkgs" (evaluated nixpkgs) to use for e.g. non-builtin fetchers
-      inherit pkgs;
-    };
-
-in
-mkSources (mkConfig {}) // { __functor = _: settings: mkSources (mkConfig settings); }
diff --git a/pkgs/blynn-compiler.nix b/pkgs/blynn-compiler.nix
deleted file mode 100644
index c9cd02d..0000000
--- a/pkgs/blynn-compiler.nix
+++ /dev/null
@@ -1,22 +0,0 @@
-{ stdenvNoCC, lib, mescc-tools-seed }:
-
-stdenvNoCC.mkDerivation {
-  name = "blynn-compiler";
-  src = lib.cleanSource ../.;
-  nativeBuildInputs = [ mescc-tools-seed ];
-
-  postPatch = ''
-    patchShebangs go.sh
-  '';
-
-  buildPhase = ''
-    ./go.sh
-  '';
-
-  installPhase = ''
-    mkdir -p $out/bin $out/share
-    cp bin/raw $out/share
-    cp bin/vm $out/bin
-    cp bin/precisely $out/bin
-  '';
-}
diff --git a/pkgs/default.nix b/pkgs/default.nix
deleted file mode 100644
index 4de0024..0000000
--- a/pkgs/default.nix
+++ /dev/null
@@ -1,5 +0,0 @@
-{ sources ? import ../nix/sources.nix, pkgs ? import sources.nix }:
-
-import sources.nixpkgs {
-  overlays = [ (import ./packages.nix) ];
-}
diff --git a/pkgs/kaem.nix b/pkgs/kaem.nix
deleted file mode 100644
index 760043c..0000000
--- a/pkgs/kaem.nix
+++ /dev/null
@@ -1,12 +0,0 @@
-{ stdenv, fetchgit }:
-
-stdenv.mkDerivation {
-  name = "kaem";
-  src = fetchgit {
-    url = "https://git.savannah.nongnu.org/git/mescc-tools.git";
-    rev = "be437350ae1589c202e824ecdb97fe9ac6c470a3 ";
-    sha256 = "0pv61cd0zw2jb09xa30vfmf22vhh2ndjdpl3g88g27c4j4dn1f19";
-  };
-  sourceDir = [ "Kaem" ];
-  makeFlags = [ "PREFIX=${placeholder "out"}" ];
-}
diff --git a/pkgs/m2-planet.nix b/pkgs/m2-planet.nix
deleted file mode 100644
index 11bd817..0000000
--- a/pkgs/m2-planet.nix
+++ /dev/null
@@ -1,15 +0,0 @@
-{ stdenv, fetchFromGitHub }:
-
-stdenv.mkDerivation rec {
-  pname = "m2-planet";
-  version = "unstable-2020-12-21";
-
-  src = fetchFromGitHub {
-    owner = "oriansj";
-    repo = pname;
-    rev = "4de11b1bc634c32c531effb32bc687aa14b602d7";
-    sha256 = "02dqbmhddk8zf615a3qsn56fq7xrkpqziw3api7pmdj8ih4k565s";
-  };
-
-  makeFlags = [ "PREFIX=${placeholder "out"}" ];
-}
diff --git a/pkgs/mes-m2.nix b/pkgs/mes-m2.nix
deleted file mode 100644
index 4c925ed..0000000
--- a/pkgs/mes-m2.nix
+++ /dev/null
@@ -1,15 +0,0 @@
-{ stdenv, fetchFromGitHub, mescc-tools }:
-
-stdenv.mkDerivation rec {
-  name = "mes-m2";
-
-  src = fetchFromGitHub {
-    owner = "oriansj";
-    repo = name;
-    rev = "6105916f9a58a879fc71f84c3df4fa622696a3ee";
-    sha256 = "0h3wivrvwz8czilpii155xxkhm15hfhawmiih248ggkanpjmkcvm";
-  };
-
-  nativeBuildInputs = [ mescc-tools ];
-  makeFlags = [ "PREFIX=${placeholder "out"}" ];
-}
diff --git a/pkgs/mescc-tools-seed.nix b/pkgs/mescc-tools-seed.nix
deleted file mode 100644
index d6a665d..0000000
--- a/pkgs/mescc-tools-seed.nix
+++ /dev/null
@@ -1,23 +0,0 @@
-{ stdenvNoCC, fetchFromGitHub }:
-
-stdenvNoCC.mkDerivation {
-  pname = "mescc-tools-seed";
-  version = "unstable-2021-01-10";
-
-  src = fetchFromGitHub {
-    owner = "oriansj";
-    repo = "mescc-tools-seed";
-    rev = "0f549aee03a6d97dd441ec17ea832aa5229f3e77";
-    sha256 = "1gw7g614rcv67wlm7rnl58ndaanl4l9rq29h2x3lqxibw9nsvz5n";
-    fetchSubmodules = true;
-  };
-
-  makeFlags = [ "PREFIX=${placeholder "out"}" ];
-
-  dontStrip = true;
-
-  installPhase = ''
-    mkdir -p $out/bin
-    cp ./bin/* $out/bin
-  '';
-}
diff --git a/pkgs/mescc-tools.nix b/pkgs/mescc-tools.nix
deleted file mode 100644
index 06707f1..0000000
--- a/pkgs/mescc-tools.nix
+++ /dev/null
@@ -1,16 +0,0 @@
-{ stdenv, fetchgit, kaem }:
-
-stdenv.mkDerivation rec {
-  pname = "mescc-tools";
-  version = "1.1.0";
-
-  src = fetchgit {
-    url = "https://git.savannah.nongnu.org/git/mescc-tools.git";
-    rev = "Release_${version}";
-    sha256 = "1asvlkrvf2kk7hn8wp6p7rmsg2p3dijk2l709j4hklyljl1zj0bp";
-  };
-
-  nativeBuildInputs = [ kaem ];
-
-  makeFlags = [ "PREFIX=${placeholder "out"}" ];
-}
diff --git a/pkgs/packages.nix b/pkgs/packages.nix
deleted file mode 100644
index b434cad..0000000
--- a/pkgs/packages.nix
+++ /dev/null
@@ -1,11 +0,0 @@
-# Bootstrappable packages as a Nixpkgs overlay
-self: super:
-
-{
-  blynn-compiler = super.callPackage ./blynn-compiler.nix { };
-  kaem = super.callPackage ./kaem.nix { };
-  m2-planet = super.callPackage ./m2-planet.nix { };
-  mes-m2 = super.callPackage ./mes-m2.nix { };
-  mescc-tools = super.callPackage ./mescc-tools.nix { };
-  mescc-tools-seed = super.callPackage ./mescc-tools-seed.nix { };
-}
diff --git a/precisely.hs b/precisely.hs
index 554858b..828304b 100644
--- a/precisely.hs
+++ b/precisely.hs
@@ -90,7 +90,8 @@ instance Integral Word where
 instance Eq Word where (==) = wordEq
 instance Ord Word where (<=) = wordLE
 
-data Word64 = Word64 Word Word deriving Eq
+data Word64 = Word64 Word Word
+instance Eq Word64 where Word64 a b == Word64 c d = a == c && b == d
 instance Ring Word64 where
   Word64 a b + Word64 c d = uncurry Word64 $ word64Add a b c d
   Word64 a b - Word64 c d = uncurry Word64 $ word64Sub a b c d
@@ -106,7 +107,8 @@ instance Ord Word64 where
     | True = b <= d
 
 -- Multiprecision arithmetic.
-data Integer = Integer Bool [Word] deriving Eq
+data Integer = Integer Bool [Word]
+instance Eq Integer where Integer xsgn xs == Integer ysgn ys = xsgn == ysgn && xs == ys
 instance Ring Integer where
   Integer xsgn xs + Integer ysgn ys
     | xsgn == ysgn = Integer xsgn $ mpAdd xs ys
@@ -212,6 +214,10 @@ class Monad m where
 liftA2 f x y = f <$> x <*> y
 (>>) f g = f >>= \_ -> g
 class Eq a where (==) :: a -> a -> Bool
+instance Eq Bool where
+  True == True = True
+  False == False = True
+  _ == _ = False
 instance Eq Int where (==) = intEq
 instance Eq Char where (==) = charEq
 ($) f x = f x
@@ -429,13 +435,13 @@ foldrWithKey f = go where
 toAscList = foldrWithKey (\k x xs -> (k,x):xs) []
 
 -- Syntax tree.
-data Type = TC String | TV String | TAp Type Type deriving Eq
+data Type = TC String | TV String | TAp Type Type
 arr a b = TAp (TAp (TC "->") a) b
 data Extra = Basic Int | Const Integer | ChrCon Char | StrCon String
 data Pat = PatLit Extra | PatVar String (Maybe Pat) | PatCon String [Pat]
 data Ast = E Extra | V String | A Ast Ast | L String Ast | Pa [([Pat], Ast)] | Ca Ast [(Pat, Ast)] | Proof Pred
 data Constr = Constr String [(String, Type)]
-data Pred = Pred String Type deriving Eq
+data Pred = Pred String Type
 data Qual = Qual [Pred] Type
 noQual = Qual []
 
@@ -444,6 +450,14 @@ typeVars = \case
   TV v -> [v]
   TAp x y -> typeVars x `union` typeVars y
 
+instance Eq Type where
+  (TC s) == (TC t) = s == t
+  (TV s) == (TV t) = s == t
+  (TAp a b) == (TAp c d) = a == c && b == d
+  _ == _ = False
+
+instance Eq Pred where (Pred s a) == (Pred t b) = s == t && a == b
+
 data Instance = Instance
   -- Type, e.g. Int for Eq Int.
   Type
@@ -813,7 +827,12 @@ addLets ls x = foldr triangle x components where
     insLams vs t = foldr L t vs
     in foldr (\(x:xt) t -> A (L x t) $ maybeFix x $ insLams xt $ suball $ maybe undefined id $ lookup x ls) (suball expr) tnames
 
-data Assoc = NAssoc | LAssoc | RAssoc deriving Eq
+data Assoc = NAssoc | LAssoc | RAssoc
+instance Eq Assoc where
+  NAssoc == NAssoc = True
+  LAssoc == LAssoc = True
+  RAssoc == RAssoc = True
+  _ == _ = False
 precOf s precTab = maybe 5 fst $ mlookup s precTab
 assocOf s precTab = maybe LAssoc snd $ mlookup s precTab
 
